[AMD] Turn on TF32 for aten::mm

xw285cornell opened the pull request:
Summary: hipblaslt supports TF32, so adding the support.

Test Plan: CI

eqy commented on aten/src/ATen/Context.cpp:
```diff
@@ -223,10 +224,26 @@ void Context::setBenchmarkLimitCuDNN(int b) {
 bool Context::allowTF32CuBLAS() const {
+#ifdef USE_ROCM
+    const auto allow_tf32 = c10::utils::get_env(hipblaslt_allow_tf32);
+    if (!allow_tf32.has_value() || allow_tf32.value() != "1") {
```
Might be simpler to use ` c10::utils::check_env(...) == true`

eqy replied:
Also consider setting to `static` if it is only expected to be set once

eqy commented on aten/src/ATen/Context.cpp:
```diff
@@ -171,6 +171,7 @@ bool Context::userEnabledOverrideableSDP() const {
 }
 
 static constexpr const auto cublas_config_var_name = "CUBLAS_WORKSPACE_CONFIG";
+static constexpr const auto hipblaslt_allow_tf32 = "HIPBLASLT_ALLOW_TF32";
```
Is there a distinction between HIPBLASLT/HIPBLAS? For CUDA backends the switch would apply to both

xw285cornell replied:
Yeah, but I don't think we care too much about hipblas because it is very very slow on MI300 and afaiu all the new development are on hipblaslt

eqy reviewed (commented):
This implementation seems a bit confusing as it appears to just discard the runtime setting/context manager if that is used? That could cause test breakages as we have many decorators like `@tf32_on_and_off` that exercise both modes at runtime.

xw285cornell commented:
@pytorchbot label "topic: not user facing"

jeffdaily commented:
@xw285cornell how goes this experiment?  Should we land it?

xw285cornell commented:
sure i'll land. We have a code freeze so I was waiting.

xw285cornell committed:
[AMD] Turn on TF32 for aten::mm (#139869)

Summary:

hipblaslt supports TF32, so adding the support.

Test Plan: CI

Reviewed By: leitian, xintwfb, yoyoyocmu

Differential Revision: D65435392

jataylo commented:
@xw285cornell, @jeffdaily @amdfaa looks like this PR caused some failures:
https://hud.pytorch.org/pytorch/pytorch/commit/e0bdae7884aed09d9e3f1a3f7a53c095e74a9aff

>2024-12-15T10:49:40.3201209Z AssertionError: 'forward call that caused the error' not found in 'torch.backends.cuda.matmul.allow_tf32 is not supported on ROCm by default. Please set environment variable HIPBLASLT_ALLOW_TF32=1 to enable it. (Triggered internally at /var/lib/jenkins/workspace/aten/src/ATen/Context.cpp:256.)'
2024-12-15T10:49:40.3202070Z

jeffdaily commented:
@pytorchbot revert -m "causing ROCm CI failures, need to investigate, revert for now" -c nosignal

jeffdaily commented:
dynamo/test_inline_inbuilt_nn_modules.py::InlineInbuiltNNModulesMiscTests::test_anomaly_aot_autograd_inline_inbuilt_nn_modules [GH job link](https://github.com/pytorch/pytorch/actions/runs/12337836654/job/34432521287) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/e0bdae7884aed09d9e3f1a3f7a53c095e74a9aff)

dynamo/test_logging.py::LoggingTests::test_logs_out [GH job link](https://github.com/pytorch/pytorch/actions/runs/12337836654/job/34432521358) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/e0bdae7884aed09d9e3f1a3f7a53c095e74a9aff)

dynamo/test_dynamic_shapes.py::DynamicShapesMiscTests::test_anomaly_aot_autograd_dynamic_shapes [GH job link](https://github.com/pytorch/pytorch/actions/runs/12337836654/job/34432521425) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/e0bdae7884aed09d9e3f1a3f7a53c095e74a9aff)

dynamo/test_misc.py::MiscTests::test_anomaly_aot_autograd [GH job link](https://github.com/pytorch/pytorch/actions/runs/12337836654/job/34432521486) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/e0bdae7884aed09d9e3f1a3f7a53c095e74a9aff)

xw285cornell commented:
I'll take a look, those tests are not running on the PR by default?

xw285cornell commented:
hmm it's a TORCH_WARN, i thought it shouldn't fail things?

xw285cornell commented:
Reland in #143549 143549
